### DISCUSSION TITLE
Replace all alpha pixels with opacity <= x with (opaque) white?

### DISCUSSION THREAD

**Discussion by vin9x** [ImageMagick/ImageMagick#5201] [Help] [answered]
Can this be done?

**Comment by snibgo** [comment]
```
magick in.ext -alpha extract -negate -threshold 34% out.png
```

**Reply by vin9x** [reply]
> ```
> magick in.ext -alpha extract -negate -threshold 34% out.png
> ```

Is it possible to manually define the color those pixels are replaced with, so that it does not necessarily have to be white?
And could you add a secondary condition that the RGB values of those pixels destined to be modified have to be 0,0,0?

Lastly, how would one make all pixels above a certain opacity fully opaque (while retaining their original RGB values)?

**Comment by snibgo** [comment]
The image from "-alpha extract -negate -threshold 34%" can be used as a mask for use in a composite operation. So we can compose one image over another with a mask, so where the mask is black the pixels come from the first image, and where it is white they come from the second. Like this (Windows BAT syntax):
```
magick ^
  in.ext ^
  ( -clone 0 -fill #abc -colorize 100 ) ^
  ( -clone 0 -alpha extract -negate -threshold 34%% ) ^
  -compose Over -composite ^
  out.ext
```
This reads a file, makes a copy and turns every pixel to colour #abc, then makes a mask, and makes the composite.

BEWARE: I haven't tested any of these commands, and may have made mistakes.

> And could you add a secondary condition that the RGB values of those pixels destined to be modified have to be 0,0,0?

You can add as many condition to the mask as you want. For example, we can make two masks and make the final mask from the lightest of those two.

```
magick ^
  in.ext ^
  ( -clone 0 -fill #abc -colorize 100 ) ^
  ( -clone 0 ^
    ( +clone -alpha extract -negate -threshold 34%% ) ^
    ( +clone -fill White +opaque Black ) ^
    -compose Lighten -composite ^
  ) ^
  -compose Over -composite ^
  out.ext
```

> Lastly, how would one make all pixels above a certain opacity fully opaque (while retaining their original RGB values)?

Make the second image:
```
  ( -clone 0 -alpha opaque ) ^
```

For this requirement, you could simply manipulate the alpha channel.

**Reply by vin9x** [reply]
Thanks, very much appreciated! Will have to think about this for a bit.

One final thing: What's the easiest way to get a readout from IM if one of those operations has actually occured? So that I can tell afterwards if there actually were rgb0,0,0 pixels within the opacity threshold that were thus converted to white/#abc?

**Comment by snibgo** [comment]
You can insert debugging writes. This is a very useful way of checking results are as you expect. For example:
```
magick ^
  in.ext ^
  ( -clone 0 -fill #abc -colorize 100 ) ^
  ( -clone 0 ^
    ( +clone -alpha extract -negate -threshold 34%% +write m0.png ) ^
    ( +clone -fill White +opaque Black +write m1.png ) ^
    -compose Lighten -composite ^
    +write m2.png ^
  ) ^
  -compose Over -composite ^
  out.ext
```
Then you can look at m0.png etc and judge whether they are as you expect.

**Reply by vin9x** [reply]
Hmm, the RGB(0,0,0) condition does not seem to work for me. It turns all pixels < x opacity into the specified color, even if they have color values other than black.

**Comment by snibgo** [comment]
Yeah, I forgot to delete the first image in the nested list.
```
magick ^
  toes_holed.png ^
  ( -clone 0 -fill #abc -colorize 100 ) ^
  ( -clone 0 ^
    ( +clone -alpha extract -negate -threshold 34%% +write m0.png ) ^
    ( +clone -fill White +opaque Black +write m1.png ) ^
    -delete 0 ^
    -compose Lighten -composite ^
    +write m2.png ^
  ) ^
  -compose Over -composite ^
  x.png
```

**Reply by vin9x** [reply]
I'm afraid this now has the opposite problem, where it does not properly convert the pixels that do fulfill the condition.
I am using the following syntax as comparison but it is too slow for practice:
`magick in.ext -channel rgba -fx "u.r==0 && u.g==0 && u.b==0 && u.a<thresh?white:u" -alpha off out.ext`
My gut tells me that the masking method simply does not work for this, otherwise the above syntax would probably figure out a way to run faster internally.

**Comment by snibgo** [comment]
What version of IM do you use, on what platform?

Please look at the debugging images m0.png, m1.png and m2.png. Are they as you expect?

Please post a sample input image.

What value are you using for the threshold?

**Reply by vin9x** [reply]
7.1.0.19
Even with threshold 240/255 it doesn't work for the transparent strip at the edge of this file: https://mega.nz/file/5MsQGJzK#B2geDT6uwNE7M-SETlqDSWu1J0vOqj6tmExN2J8iEO8
The single line syntax from above works and makes the strip white but it's incredibly slow.
I'm not entirely sure what the debugging images are supposed to look like.

**Comment by snibgo** [comment, marked as answer]
```
magick in.ext -channel rgba -fx "u.r==0 && u.g==0 && u.b==0 && u.a<thresh?white:u" -alpha off out.ext
```
That won't work. The syntax is bad. What value are you using instead of "thresh"?

> The single line syntax from above works and makes the strip white but it's incredibly slow.

Please quantify "incredibly slow".

This takes about 0.4 second:
```
magick ^
  logocircuitocartegena.dds ^
  ( -clone 0 -fill #fff -colorize 100 ) ^
  ( -clone 0 ^
    ( +clone -alpha extract -negate -threshold 34%% ) ^
    ( +clone -fill White +opaque Black ) ^
    -delete 0 ^
    -compose Lighten -composite ^
  ) ^
  -alpha off ^
  -compose Over -composite ^
  x.png
```

This takes about 0.3 seconds:
```
magick logocircuitocartegena.dds -channel rgba -fx "u.r==0 && u.g==0 && u.b==0 && u.a<0.34?white:u" -alpha off f.png
```
Each "u." is redundant, and the speed slightly increases if we remove them. This takes about 0.2 seconds:
```
magick logocircuitocartegena.dds -channel rgba -fx "r==0 && g==0 && b==0 && a<0.34?white:u" -alpha off f2.png
```
The outputs x.png, f.png and f2.png are identical.

**Reply by vin9x** [reply]
Something must be messed up with my operating system. I am using the exact syntax you posted but the mask methoed still produces a wrong result, while the other one is correct but takes multiple seconds.

**Comment by snibgo** [comment]
Your IM version, 7.1.0-19, is before the performance of "-fx" greatly improved. I suggest you upgrade to the current version.

You still haven't said what platform you are on. Bash, or Windows BAT, or what?

**Reply by vin9x** [reply]
Windows BAT,
Is there are portable version available with the improved performance?

**Comment by snibgo** [comment]
https://imagemagick.org/script/download.php#windows

**Reply by vin9x** [reply]
The mask still doesn't work but at least with the newest version the other syntax works reasonably fast.
So thanks!
